rpc/RPCTransaction.java

```diff
@@ -170,38 +174,23 @@ public void onNext(TransactionProto.Transaction.Res res) {
                 final UUID requestId = UUID.fromString(res.getId());
                 final ResponseCollector collector = collectors.get(requestId);
                 assert collector != null : UNKNOWN_REQUEST_ID.message(requestId);
-                collector.add(new Response.Ok(res));
+                collector.add(new Response.Result(res));
             }
 
             @Override
             public void onError(Throwable ex) {
                 assert ex instanceof StatusRuntimeException : "The server sent an exception of unexpected type " + ex;
                 // TODO: this isn't nice - an error from one request isn't really appropriate for all of them (see #180)
-                closeError((StatusRuntimeException) ex);
+                close(new Response.Done.Error((StatusRuntimeException) ex));
             }
 
             @Override
             public void onCompleted() {
-                closeSuccess();
-            }
-
-            private void closeError(StatusRuntimeException ex) {
-                if (isOpen.compareAndSet(true, false)) {
-                    collectors.error(new Response.Error(ex));
-                    collectors.clear();
-                    requestObserver.onCompleted();
-                }
+                close(new Response.Done.Completed());
```

Comment:
Here is the main bug fix. Previously we use `closeSuccess()` which doesn't post an error response to all pending requests, which means they may still be blocked waiting for server response forever. Now we make sure we always post the error response (implemented in `close` method) so that all pending requests will be unblocked.